File: LC_JS-SOLUTIONS/234. Palindrome Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function (head) {
  // Funciton to reverse the second half

  function reverseSecHalf(head) {
    let prev = null;
    current = head;
    while (current) {
      let nextNode = current.next;
      current.next = prev;
      prev = current;
      current = nextNode;
    }
    return prev;
  }

  // Inital get the half of the list

  let slow = head;
  let fast = head;

  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;
  }

  // Get the reverse of the 2nd half as slow is pointed to 2nd half

  let revSecHalf = reverseSecHalf(slow);

  // Check the palindrome or not
  let firstHalf = head;

  while (revSecHalf) {
    if (revSecHalf.val != firstHalf.val) {
      return false;
    }
    firstHalf = firstHalf.next;
    revSecHalf = revSecHalf.next;
  }
  return true;
};

File: LC-Python-Solution/234. Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:

        def reverseSecHalf(head):
            prev = None
            current = head
            while current:
                nextNode = current.next
                current.next = prev
                prev = current
                current = nextNode
            return prev
            
        # first get the half of list
        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        # second reverse the second half as slow point to 2nd half onward
        revSecHalf = reverseSecHalf(slow)

        # now check if they are palindrome or not
        firstHalf =  head
        while revSecHalf:
            if revSecHalf.val != firstHalf.val:
                return False
            
            revSecHalf = revSecHalf.next
            firstHalf = firstHalf.next
        
        return True

        
        
        


        